idaes/core/util/scaling.py

```diff
@@ -1041,3 +1047,292 @@ def set_derivative_factor_from_state(self, deriv, nominal_wrt=1.0):
+        EXPR.NPV_ProductExpression: _get_nominal_value_for_product,
+        EXPR.DivisionExpression: _get_nominal_value_for_division,
+        EXPR.NPV_DivisionExpression: _get_nominal_value_for_division,
+        EXPR.PowExpression: _get_nominal_value_for_power,
+        EXPR.NPV_PowExpression: _get_nominal_value_for_power,
+        EXPR.NegationExpression: _get_nominal_value_negation,
+        EXPR.NPV_NegationExpression: _get_nominal_value_negation,
+        EXPR.AbsExpression: _get_nominal_value_abs,
+        EXPR.NPV_AbsExpression: _get_nominal_value_abs,
+        EXPR.UnaryFunctionExpression: _get_nominal_value_for_unary_function,
+        EXPR.NPV_UnaryFunctionExpression: _get_nominal_value_for_unary_function,
+        EXPR.Expr_ifExpression: _get_nominal_value_expr_if,
+        EXPR.ExternalFunctionExpression: _get_nominal_value_external_function,
+        EXPR.NPV_ExternalFunctionExpression: _get_nominal_value_external_function,
+        EXPR.LinearExpression: _get_nominal_value_for_sum,
+    }
+
+    def exitNode(self, node, data):
+        """Callback for :class:`pyomo.core.current.StreamBasedExpressionVisitor`. This
+        method is called when moving back up the tree in a depth first search."""
+
+        # first check if the node is a leaf
+        nodetype = type(node)
+
+        # TODO: What do we do with an indexed component? Fail outright?
+
+        if nodetype in native_types or nodetype in pyomo_constant_types:
+            if node == 0:
+                return [1]  # Catch cases of zero and use default scaling of 1
+            return [node]
+
+        node_func = self.node_type_method_map.get(nodetype, None)
+        if node_func is not None:
+            return node_func(self, node, data)
+
+        elif not node.is_expression_type():
+            # this is a leaf, but not a native type
+            if nodetype is _PyomoUnit:
+                return [1]
+            else:
+                # might want to add other common types here
+                sf = get_scaling_factor(node, default=1, warning=self.warning)
+                return [1 / sf]
+
+        # not a leaf - check if it is a named expression
+        if (
+            hasattr(node, "is_named_expression_type")
+            and node.is_named_expression_type()
+        ):
+            return self._get_nominal_value_single_child(node, data)
+
+        raise TypeError(
+            "An unhandled expression node type: {} was encountered while retrieving the"
+            " scaling factor of expression".format(str(nodetype), str(node))
+        )
+
+
+def set_constraint_scaling_max_magnitude(
+    component, warning: bool = True, overwrite: bool = False, descend_into: bool = True
+):
+    """
+    Set scaling factors for constraints using maximum expected magnitude of additive terms in expression.
+    Scaling factor for constraints will be 1 / max(abs(nominal value)).
+
+    Args:
+        component: a Pyomo component to set constraint scaling factors for.
```

Comment:
This section flips back and forth between "component" and "constraint". We'd generally prefer `Constraint` scaling to happen with `constraint_scaling_transform` rather than `set_scaling_factor`, because `constraint_scaling_transform` also affects the size of the primal residual when IPOPT evaluates how close to satisfied a constraint is, while `set_scaling_factor` affects only the matrix inversion.

Reply:
No, this was built explicitly as the first step to replace `constraint_scaling_transform` and to transposition instead to using the Pyomo scaling transformation (and to not pass any suffixes to IPOPT).

Also, the use of component and constraint is deliberate- the method can take Pyomo Blocks, in which case it will scale all the constraints within the Block.